### PR TITLE
harden tests for heap graph reconstruction

### PR DESCRIPTION
**Pull request by sam-herman**
# Description

- Harden test for the cases where heap graph reconstruction and ravv to graph ordinal mappings are not the identity mapping.
- Fix bug when graph ord mapping is used with the remmaped RAVV
- Move to remmaped interface for buildAndMerge to avoid bugs of remmpaing ords twice
- Add interface for providing the executor pools for the `buildAndMerge` method  -important for downstream dependencies with centralized resource management

### CODE REVIEW

File: jvector-base/src/main/java/io/github/jbellis/jvector/graph/GraphIndexBuilder.java [hunk 1]
```diff
@@ -439,6 +439,14 @@ public static GraphIndexBuilder rescore(GraphIndexBuilder other, BuildScoreProvi
     }
 
     public ImmutableGraphIndex build(RandomAccessVectorValues ravv) {
+        return reallyBuild(ravv);
+    }
+
+    public ImmutableGraphIndex build(RandomAccessVectorValues ravv, int[] graphToRavvOrdMap) {
```
Comment (marianotepper) [inline jvector-base/src/main/java/io/github/jbellis/jvector/graph/GraphIndexBuilder.java]: We should not add this method here. It should be the responsibility of the caller to use the remapped RAVV, not a conversion that we want done by JV. Just like the change in buildAndMergeNewNodes

File: jvector-base/src/main/java/io/github/jbellis/jvector/graph/GraphIndexBuilder.java [hunk 2]
```diff
@@ -990,15 +996,47 @@ private void loadV3(RandomAccessReader in, int size) throws IOException {
...
+     * @param addHierarchy whether to add hierarchical structures while building the graph.
+     * @param simdExecutor the ForkJoinPool executor used for SIMD tasks during graph building.
+     * @param parallelExecutor the ForkJoinPool executor used for general parallelization during graph building.
+     *
+     * @return the in-memory representation of the graph index.
+     * @throws IOException if an I/O error occurs during the graph loading or conversion process.
+     */
+    @Experimental
+    public static ImmutableGraphIndex buildAndMergeNewNodes(RandomAccessReader in,
+                                                            RemappedRandomAccessVectorValues newVectors,
```
Comment (marianotepper) [inline jvector-base/src/main/java/io/github/jbellis/jvector/graph/GraphIndexBuilder.java:1022]: This method should receive RandomAccessVectorValues isntead of RemappedRandomAccessVectorValues

File: jvector-tests/src/test/java/io/github/jbellis/jvector/graph/OnHeapGraphIndexTest.java [hunk 3]
```diff
@@ -166,12 +202,51 @@ public void testReconstructionOfOnHeapGraphIndex() throws IOException {
...
     }
 
+    /**
+     * Create an {@link OnHeapGraphIndex} persist it as a {@link OnDiskGraphIndex} and reconstruct back to a mutable {@link OnHeapGraphIndex}
+     * Using random mapping from graph node id to ravv ordinal
+     * Make sure that both graphs are equivalent
+     * @throws IOException
+     */
+    @Test
+    public void testReconstructionOfOnHeapGraphIndex_withNonIdentityOrdinalMapping() throws IOException {
```
Comment (marianotepper) [inline jvector-tests/src/test/java/io/github/jbellis/jvector/graph/OnHeapGraphIndexTest.java:213]: Remove underscore in name, use camel case

File: jvector-tests/src/test/java/io/github/jbellis/jvector/graph/OnHeapGraphIndexTest.java [hunk 4]
```diff
@@ -129,14 +141,38 @@ public void tearDown() {
...
 
     /**
      * Create an {@link OnHeapGraphIndex} persist it as a {@link OnDiskGraphIndex} and reconstruct back to a mutable {@link OnHeapGraphIndex}
+     * Using identity mapping from graph node id to ravv ordinal
      * Make sure that both graphs are equivalent
      * @throws IOException
      */
     @Test
-    public void testReconstructionOfOnHeapGraphIndex() throws IOException {
+    public void testReconstructionOfOnHeapGraphIndex_withIdentityOrdinalMapping() throws IOException {
```
Comment (marianotepper) [inline jvector-tests/src/test/java/io/github/jbellis/jvector/graph/OnHeapGraphIndexTest.java:176]: Remove underscore in name, use camel case

File: jvector-tests/src/test/java/io/github/jbellis/jvector/graph/OnHeapGraphIndexTest.java [hunk 5]
```diff
@@ -166,12 +202,51 @@ public void testReconstructionOfOnHeapGraphIndex() throws IOException {
...
+        }
+    }
+
     /**
      * Create {@link OnDiskGraphIndex} then append to it via {@link GraphIndexBuilder#buildAndMergeNewNodes}
      * Verify that the resulting OnHeapGraphIndex is equivalent to the graph that would have been alternatively generated by bulk index into a new {@link OnDiskGraphIndex}
      */
     @Test
-    public void testIncrementalInsertionFromOnDiskIndex() throws IOException {
+    public void testIncrementalInsertionFromOnDiskIndex_withIdentityOrdinalMapping() throws IOException {
```
Comment (marianotepper) [inline jvector-tests/src/test/java/io/github/jbellis/jvector/graph/OnHeapGraphIndexTest.java:250]: Remove underscore in name, use camel case

File: jvector-base/src/main/java/io/github/jbellis/jvector/graph/GraphIndexBuilder.java (hunk 1 above)
Comment (sam-herman) [reply jvector-base/src/main/java/io/github/jbellis/jvector/graph/GraphIndexBuilder.java]: Agree, will remove it!

File: jvector-base/src/main/java/io/github/jbellis/jvector/graph/GraphIndexBuilder.java (hunk 2 above)
Comment (sam-herman) [reply jvector-base/src/main/java/io/github/jbellis/jvector/graph/GraphIndexBuilder.java:1022]: I think for this one we always take a `RemappedRandomAccessVectorValues` because we have to assume some ordinals between graph nodeIds and RAVV might not fit (remember that the graph was built with different RAVV instance), therefore to prevent user error we force remmaped type. We can use identity mapping in the edge case that there is not difference, but that is highly unlikely.

File: jvector-tests/src/test/java/io/github/jbellis/jvector/graph/OnHeapGraphIndexTest.java (hunk 4 above)
Comment (sam-herman) [reply jvector-tests/src/test/java/io/github/jbellis/jvector/graph/OnHeapGraphIndexTest.java:176]: In Java for testing there is a naming convention of `[MethodUnderTest]_[Scenario]_[ExpectedResult]` for unit tests.
It's especially common when branching several scenarios under the same method being tested. With that being said I don't mind changing to whatever you feel is more readable :)

Comment (marianotepper) [reply jvector-tests/src/test/java/io/github/jbellis/jvector/graph/OnHeapGraphIndexTest.java:176]: I did not know that. I learned something :-)
Ok with me then!
